mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMemorySearchHighlighter.java

```diff
@@ -0,0 +1,190 @@
+import org.apache.lucene.store.Directory;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.collect.ImmutableSet;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class LuceneMemorySearchHighlighter implements SearchHighlighter {
+
+    private final LuceneMessageSearchIndex luceneMessageSearchIndex;
+    private final Analyzer analyzer;
+    private final Formatter formatter;
+    private final Fragmenter fragmenter;
+    private final IndexSearcher searcher;
+
+    private boolean suffixMatch = false;
+
+    public LuceneMemorySearchHighlighter(LuceneMessageSearchIndex luceneMessageSearchIndex,
+                                         SearchHighlighterConfiguration searchHighlighterConfiguration,
+                                         IndexWriter indexWriter) throws IOException {
+        this.luceneMessageSearchIndex = luceneMessageSearchIndex;
+        this.analyzer = new StandardAnalyzer();
+
+        DirectoryReader directoryReader = DirectoryReader.open(indexWriter);
+        this.searcher = new IndexSearcher(directoryReader);
+
+        this.formatter = new SimpleHTMLFormatter(searchHighlighterConfiguration.preTagFormatter(), searchHighlighterConfiguration.postTagFormatter());
+        this.fragmenter = new SimpleFragmenter(searchHighlighterConfiguration.fragmentSize());
+    }
+
+    @Override
+    public Flux<SearchSnippet> highlightSearch(MultimailboxesSearchQuery expression, MailboxSession session, long limit) {
+        // Todo check if the mailbox is accessible
+
+        SearchQuery searchQuery = expression.getSearchQuery();
+        ImmutableSet<MailboxId> inMailboxes = expression.getInMailboxes();
+        return Mono.fromCallable(() -> luceneMessageSearchIndex.searchDocument(inMailboxes, searchQuery, (int) limit)).flatMapMany(Flux::fromIterable)
+            .map(document -> Throwing.supplier(() -> buildSearchSnippet(document, searchQuery)).get());
+    }
+
+
+    public List<SearchSnippet> highlightSearchDirectly(MultimailboxesSearchQuery expression, MailboxSession session, long limit) {
+        BooleanQuery.Builder queryBuilder = new BooleanQuery.Builder();
+        //queryBuilder.add(buildQueryFromMailboxes(expression.getInMailboxes()), BooleanClause.Occur.MUST);
+        List<SearchQuery.Criterion> crits = expression.getSearchQuery().getCriteria();
+
+        if (crits.getFirst() instanceof SearchQuery.TextCriterion textCriterion) {
+            String searchText = textCriterion.getOperator().getValue().toUpperCase(Locale.US);
+
+            BooleanQuery subjectAndBodyQuery = new BooleanQuery.Builder()
+                .add(createTermQuery(BODY_FIELD, searchText), BooleanClause.Occur.SHOULD)
+                .add(createTermQuery(BASE_SUBJECT_FIELD, searchText), BooleanClause.Occur.SHOULD)
+                .build();
+            queryBuilder.add(subjectAndBodyQuery, BooleanClause.Occur.MUST);
+        }
+
+        BooleanQuery query = queryBuilder.build();
```

Comment:
Too much duplicated code. 

See https://github.com/apache/james-project/pull/2429/commits/e93b5f9a73b6fea7f5e00664fb2ae1ec81787bbf#r1782647302 which proves we do have too much coupling between the highlighter and the index.

I would really appreciate we ind a way to get the query from the index and not craft an add-hoc one in here.

Reply:
removed duplicate code

Ref: https://github.com/apache/james-project/pull/2429#discussion_r1782556792